### PR TITLE
New Metric: Dunn Index

### PR DESCRIPTION
**Pull request by matsumotosan**
## What does this PR do?

Fixes part of #2003

<details>
  <summary>Before submitting</summary>

- [ ] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to **update the docs**?
- [ ] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

----
:books: Documentation preview :books:: https://torchmetrics--2049.org.readthedocs.build/en/2049/

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -20,6 +20,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Added `NormalizedMutualInfoScore` metric to cluster package ([#2029](https://github.com/Lightning-AI/torchmetrics/pull/2029)
 
 
+- Added `DunnIndex` metric to cluster package ([#200]())
```
Comment (SkafteNicki) [inline CHANGELOG.md:23]:
```suggestion
- Added `DunnIndex` metric to cluster package ([#2049](https://github.com/Lightning-AI/torchmetrics/pull/2049))
```

File: docs/source/clustering/dunn_index.rst [hunk 2]
```diff
@@ -0,0 +1,21 @@
+.. customcarditem::
+   :header: Dunn Index
+   :image: https://pl-flash-data.s3.amazonaws.com/assets/thumbnails/clustering.svg
```
Comment (SkafteNicki) [inline docs/source/clustering/dunn_index.rst:3]:
```suggestion
   :image: https://pl-flash-data.s3.amazonaws.com/assets/thumbnails/default.svg
```
the `clustering.svg` does not exist, we create one in the future and change them

File: src/torchmetrics/clustering/dunn_index.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    .. math::
+        DI_m = \frac{\min_{1\leq i<j\leq m} \delta(C_i,C_j)}{\max_{1\leq k\leq m} \Delta_k}
+
+    Where :math:`C_i` is a cluster of tensors, :math:`C_j` is a cluster of tensors,
+    and :math:`\delta(C_i,C_j)` is the intercluster distance metric for :math:`m` clusters.
+
+    As input to ``forward`` and ``update`` the metric accepts the following input:
+
+    - ``preds`` (:class:`~torch.Tensor`): single integer tensor with shape ``(N,)`` with predicted cluster labels
+    - ``target`` (:class:`~torch.Tensor`): single integer tensor with shape ``(N,)`` with ground truth cluster labels
```
Comment (SkafteNicki) [inline src/torchmetrics/clustering/dunn_index.py:40]:
```suggestion
    - ``data`` (:class:`~torch.Tensor`): float tensor with shape ``(N,d)`` with the embedded data. ``d`` is the dimensionality of the embedding space.
    - ``labels`` (:class:`~torch.Tensor`): single integer tensor with shape ``(N,)`` with cluster labels
```
In PR https://github.com/Lightning-AI/torchmetrics/pull/2036 i have called it `data, labels`. I am not sure what the right naming is, but I think we should not use `preds, target` here because it indicates that there is ground truth value which there is not.

File: src/torchmetrics/clustering/dunn_index.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        self.add_state("labels", default=[], dist_reduce_fx="cat")
+
+    def update(self, x: Tensor, labels: Tensor) -> None:
+        """Update state with predictions and targets."""
+        self.x.append(x)
+        self.labels.append(labels)
+
+    def compute(self) -> Tensor:
+        """Compute mutual information over state."""
+        return dunn_index(dim_zero_cat(self.x), dim_zero_cat(self.labels))
```
Comment (SkafteNicki) [inline src/torchmetrics/clustering/dunn_index.py:81]:
```suggestion
        self.add_state("data", default=[], dist_reduce_fx="cat")
        self.add_state("labels", default=[], dist_reduce_fx="cat")

    def update(self, data: Tensor, labels: Tensor) -> None:
        """Update state with predictions and targets."""
        self.data.append(data)
        self.labels.append(labels)

    def compute(self) -> Tensor:
        """Compute mutual information over state."""
        return dunn_index(dim_zero_cat(self.data), dim_zero_cat(self.labels))
```

File: src/torchmetrics/functional/clustering/dunn_index.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        max_intracluster_distance: max intracluster distances
+
+    Returns:
+        dunn_index: Dunn index
+
+    """
+    return intercluster_distance.min() / max_intracluster_distance.max()
+
+
+def dunn_index(x: Tensor, labels: Tensor, p: float = 2) -> Tensor:
```
Comment (SkafteNicki) [inline src/torchmetrics/functional/clustering/dunn_index.py:63]:
```suggestion
def dunn_index(data: Tensor, labels: Tensor, p: float = 2) -> Tensor:
```

File: src/torchmetrics/functional/clustering/dunn_index.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    """Compute the Dunn index.
+
+    Args:
+        x: feature vectors
```
Comment (SkafteNicki) [inline src/torchmetrics/functional/clustering/dunn_index.py:67]:
```suggestion
        data: feature vectors
```

File: src/torchmetrics/functional/clustering/dunn_index.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+    Example:
+        >>> from torchmetrics.functional.clustering import dunn_index
+        >>> x = torch.tensor([0, 3, 2, 2, 1])
+        >>> labels = torch.tensor([1, 3, 2, 0, 1])
+        >>> dunn_index(preds, target)
+        tensor(1.0)
+
+    """
+    pairwise_distance, max_distance = _dunn_index_update(x, labels, p)
```
Comment (SkafteNicki) [inline src/torchmetrics/functional/clustering/dunn_index.py:82]:
```suggestion
    pairwise_distance, max_distance = _dunn_index_update(data, labels, p)
```

**Review by SkafteNicki — APPROVED** [review]

File: src/torchmetrics/functional/clustering/utils.py [hunk 8]
```diff
@@ -151,6 +151,18 @@ def calculate_contingency_matrix(
     return contingency
 
 
+def _is_real_discrete_label(x: Tensor) -> bool:
+    """Check if tensor of labels is real and discrete.
+
+    Args:
+        x: tensor
+
+    """
```
Comment (Borda) [inline src/torchmetrics/functional/clustering/utils.py:160]:
```suggestion
    """Check if tensor of labels is real and discrete.
    """
```

File: tests/unittests/clustering/test_dunn_index.py [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+    _single_target_intrinsic1,
+    _single_target_intrinsic2,
+)
+from unittests.helpers import seed_all
+from unittests.helpers.testers import MetricTester
+
+seed_all(42)
+
+
+def _np_dunn_index(x, labels, p):
```
Comment (Borda) [inline tests/unittests/clustering/test_dunn_index.py:32]:
```suggestion
def _np_dunn_index(data, labels, ord):
```
lets be more descriptive

**Review by Borda — APPROVED** [review]

File: tests/unittests/clustering/test_dunn_index.py (hunk 9 above)
Comment (matsumotosan) [reply tests/unittests/clustering/test_dunn_index.py]: Had to switch `ord` back to `p` since `ord` is a python builtin.

File: src/torchmetrics/functional/clustering/dunn_index.py [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from itertools import combinations
+from typing import Tuple
+
+import torch
+from torch import Tensor
+
+
+def _dunn_index_update(data: Tensor, labels: Tensor, p: float) -> Tuple[Tensor, Tensor]:
```
Comment (justusschock) [inline src/torchmetrics/functional/clustering/dunn_index.py:21]: nit: can we maybe rename this function? Currently it looks like this should be called in the class-based metric as update, but it isn't. Just for clarity there :)

**Review by justusschock — APPROVED** [review]
